src/filesystem.cpp

```diff
@@ -857,45 +824,77 @@ size_t LocalPath::getLeafnameByteIndex(const FileSystemAccess& fsaccess) const
 bool LocalPath::backEqual(size_t bytePos, const LocalPath& compareTo) const
 {
     auto n = compareTo.localpath.size();
-    return bytePos + n == localpath.size() && !memcmp(compareTo.localpath.data(), localpath.data() + bytePos, n);
+    return bytePos + n == localpath.size() && !localpath.compare(bytePos, n, compareTo.localpath);
 }
 
 LocalPath LocalPath::subpathFrom(size_t bytePos) const
 {
-    return LocalPath::fromLocalname(localpath.substr(bytePos));
+    LocalPath result;
+    result.localpath = localpath.substr(bytePos);
+    return result;
 }
 
-
 void LocalPath::ensureWinExtendedPathLenPrefix()
 {
 #if defined(_WIN32) && !defined(WINDOWS_PHONE)
-    localpath.append("", 1);
-    if (!PathIsRelativeW((LPWSTR)localpath.c_str()) && ((localpath.size() < 4) || memcmp(localpath.data(), L"\\\\", 4)))
-        localpath.insert(0, (const char*)(const wchar_t*)L"\\\\?\\", 8);
-    localpath.resize(localpath.size() - 1);
+    if (!PathIsRelativeW(localpath.c_str()) && ((localpath.size() < 2) || memcmp(localpath.data(), L"\\\\", 4)))
+        localpath.insert(0, L"\\\\?\\", 4);
 #endif
 }
 
-string LocalPath::substrTo(size_t bytePos) const
+LocalPath LocalPath::subpathTo(size_t bytePos) const
+{
+    LocalPath p;
+    p.localpath = localpath.substr(0, bytePos);
+    return p;
+}
+
+
+LocalPath LocalPath::insertFilenameCounter(unsigned counter, const FileSystemAccess& fsaccess)
 {
-    return localpath.substr(0, bytePos);
+    // the destination path isn't synced, save with a (x) suffix
+
+    size_t dotindex = localpath.find_last_of('.');
+    size_t sepindex = localpath.find_last_of(fsaccess.localseparator);
+
+    LocalPath result, extension;
+
+    if (dotindex == string::npos || (sepindex != string::npos && sepindex > dotindex))
+    {
+        result.localpath = localpath;
+    }
+    else
+    {
+        result.localpath = localpath.substr(0, dotindex);
+        extension.localpath = localpath.substr(dotindex);
+    }
+
+    ostringstream oss;
+    oss << " (" << counter << ")";
+
+    result.localpath += LocalPath::fromPath(oss.str(), fsaccess).localpath + extension.localpath;
+    return result;
 }
 
+
 string LocalPath::toPath(const FileSystemAccess& fsaccess) const
 {
     string path;
-    fsaccess.local2path(const_cast<string*>(&localpath), &path); // todo: const correctness for local2path etc
+#if defined(_WIN32)
+    fsaccess.local2path(const_cast<std::wstring*>(&localpath), &path);
+#else
+    fsaccess.local2path(const_cast<string*>(&localpath), &path);
```

Comment:
castings are no longer required If I'm not mistaken

Reply:
good point, which also means we don't need the ifdef - updated